### ISSUE TITLE
embed_positions of hf version of Structured Prompting ()

### ISSUE THREAD

**Issue by amurtadha** [microsoft/LMOps#54]
For the hf version of Structured Prompting (modeling_opt.py):
let:

config.max_position_embeddings=2048
 config. hidden_size=768

self.embed_positions = OPTLearnedPositionalEmbedding(config.max_position_embeddings, config.hidden_size)

past_key_values_length = 3000
attention_mask.size() is (1, 3400)

pos_embeds = self.embed_positions(attention_mask, past_key_values_length)

Now both past_key_values_length and attention_mask exceed the  max_position_embeddings. In this situation, what is the solution thank you

**Comment by sunyt32** [comment]
To utilize Structured Prompting in other Transformer designs, we should modify the corresponding position embedding function. For OPT's absolute position embedding, since the past_key_value is parallel, the real final position is not ```attention_mask.shape[1]```. The original code is:
```
positions = (torch.cumsum(attention_mask, dim=1).type_as(attention_mask) * attention_mask).long() - 1
# cut positions if `past_key_values_length` is > 0
positions = positions[:, past_key_values_length:]
```
I suggest modifying it like this:
```
start_position = past_key_values_length / chunk_num
end_position = start_position + attention_mask.shape[1] - past_key_values_length
positions = torch.arange(start_position, end_position).to(attention_mask)
```
It's just a pseudo code, you can integrate it into your own pipeline.

**Comment by amurtadha** [comment]
Thank you very much. I got the idea. However, when I test with num_shot=500 for sst2 the next_token is always Sent, while with 200 it works fine.  I just update the OPTAttention()

the orginal is :
        if attn_weights.dtype == torch.float16:
            attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(torch.float16)
        else:
           attn_weights = nn.functional.softmax(attn_weights, dim=-1)
I updated to :

 input_dtype = attn_weights.dtype
if prefix_parallel is not None and prefix_parallel > 1:
                attn_weights= attn_weights.view(bsz, self.num_heads, tgt_len, src_len)
                past_key_values_length = layer_past[0].shape[1]

                attn_weights = torch.cat((attn_weights, attn_weights[:, :, :, past_key_values_length:].repeat(1, 1, 1, (
                            prefix_parallel - 1))), dim=-1)
                attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(input_dtype)

                attn_weights = torch.cat((attn_weights[:, :, :, :past_key_values_length], \
                                          attn_weights[:, :, :, past_key_values_length:].view(attn_weights.shape[0],
                                                                                              attn_weights.shape[1],
                                                                                              attn_weights.shape[2],
                                                                                              prefix_parallel, -1).sum(
                                              dim=-2)), dim=-1)

                attn_weights= attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
else:
                attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(input_dtype)


any suggestion please

**Comment by amurtadha** [comment]
@sunyt32 it is fixed already, 
past_key_values_length = layer_past[0].shape[2]
thanks
